### DISCUSSION TITLE
Using CIP path

### DISCUSSION THREAD

**Discussion by BmorgCoding** [ottowayi/pycomm3#214] [Q&A]
I'm struggling to use the CIP path to read a PLC.

I don't believe it's an issue with pycomm3, just something I'm not doing quite right. 
PLC:
![image](https://user-images.githubusercontent.com/92693575/162208165-cd41722e-e9d0-4f09-ab80-8c8043b486f2.png)

RSLinx Who active:

![image](https://user-images.githubusercontent.com/92693575/162206854-39f6d06f-c1ee-4d67-8d21-965b3efa939b.png)

Here's the path (Shown at the bottom of who active) in RSLinx.
Here's what I tried in python:

After using the Logix driver trying to read a tag wasn't working for me, I figured I would first use the example on documents page to see if I can get a response from the PLC:

`
       imports = True
       if imports:
            from pycomm3 import (CIPDriver, Services, ClassCode,  FileObjectServices, FileObjectInstances,
                       FileObjectInstanceAttributes, Struct, UDINT, USINT, n_bytes, LogixDriver, SLCDriver)
  
  
       definitions = True
       if definitions:
           def read_singleLogixValue(PLC_IP, PLC_Tag):
               #try:
               #    print('Test')
               #    with LogixDriver(PLC_IP) as plc:
               #        read = plc.read(PLC_Tag)
  
               #        print(plc)
               #        print(read)
               #except Exception as e:
               #       print(str(e))
               with CIPDriver(PLC_IP) as plc:
                   print(plc)
             
       if __name__ == "__main__" :
  
           PLCIP = '10.32.200.59/bp/2/enet/10.32.210.32/bp/0'
           PLCType = 'Logix'
           PLCTag = 'PQDATA[28]'
  
           while True:
               if PLCType == 'Logix':
                   read_singleLogixValue(PLCIP, PLCTag)   
               time.sleep(5)   
`

Returns:
![image](https://user-images.githubusercontent.com/92693575/162207796-04c8b8da-bb89-4bec-a965-f974569011c2.png)

In the RSLinx the "A" is some sort of ethernet switch, or branch. So I replaced A in the path with enet. Tried using Backplane, backplane, bp, \ and / and couldn't quite figure out a combination that will let me connect to the PLC.
Have also tried using LogixDriver with the same path and a PLC tag to do a read, which when running the Python code freezes, doesn't give any errors or results.

My guess is I'm not building the CIP path correctly.

**Comment by ottowayi** [comment]
Hmm... I'm not sure what the issue could be, your path looks correct.  `'A'` is the label for the front port on the EN2T, so `enet` is the correct value or you can just use `2`, but there isn't an alias for `A` (yet). I think the CIP driver is actually missing the identity info, that's why when you print it you get `Device: None...`.  I have it on my list of things to fix.  If you're trying to read tags, you'll need the Logix driver not the CIP one, so we'll want to get that working.  Can you enable logging and send it?  Look [here](https://docs.pycomm3.dev/en/latest/getting_started.html#logging) for instructions, we'll want the verbose level so we can see the actual traffic back and forth.  If you're worried about posting it publicly, you can email me it instead. 

One minor change I would make is to move your `read_singleLogixValue` function out from underneath the `if` like:

```
def read_singleLogixValue():
    ...

if __name__ == "__main__" :
      ....
       while True:
           if PLCType == 'Logix':
               if definitions:
                   read_singleLogixValue(PLCIP, PLCTag)   
                   time.sleep(5)   

```

or something similar, and maybe instead of being in a `while` just have it do it once until we get it working.

**Comment by BmorgCoding** [comment]
Thanks for the response!

Will make the suggested changes and try it again. The IF statement is kind of useless code I have in place just for making it easier to navigate in visual studio, many definitions I can easily minimize with one click. But will separate it all until have something working. 

Won't actually be using CIPDriver for the code, but commented out the LogixDriver and was trying to use CIP to troubleshoot and see if my path was correct. Once I can get connected will no longer need it.

If I remove the A from the path and use only the 2 it did not work either.

Thanks for the work on the library. The large majority of the devices I connect to don't require the CIP route and pycomm is a pleasure to use.
